pkg/controllers/mcs/service_import_controller.go

```diff
@@ -0,0 +1,595 @@
+	defer logger.Info("Stopping controller")
+
+	go c.endpointSliceInformer.Informer().Run(ctx.Done())
+
+	if !cache.WaitForNamedCacheSync(ServiceImportControllerName, ctx.Done(), c.HasSynced) {
+		logger.Error(nil, "Timed out waiting for caches to sync")
+		return
+	}
+	logger.Info("Caches are synced")
+	c.worker.Run(ctx)
+	<-ctx.Done()
+}
+
+func (c *ServiceImportController) reconcile(ctx context.Context, qualifiedName common.QualifiedName) (status worker.Result) {
+	key := qualifiedName.String()
+	ctx, keyedLogger := logging.InjectLoggerValues(ctx, "control-loop", "reconcile", "object", key)
+
+	fedObject, err := fedobjectadapters.GetFromLister(
+		c.fedObjectInformer.Lister(),
+		nil,
+		qualifiedName.Namespace,
+		qualifiedName.Name,
+	)
+	if err != nil && !apierrors.IsNotFound(err) {
+		keyedLogger.Error(err, "Failed to get federated object from store")
+		return worker.StatusError
+	}
+
+	if fedObject == nil || fedObject.GetDeletionTimestamp() != nil ||
+		fedObject.GetAnnotations()[common.DryRunAnnotation] == common.AnnotationValueTrue {
+		return worker.StatusAllOK
+	}
+
+	template, err := fedObject.GetSpec().GetTemplateAsUnstructured()
+	if err != nil {
+		keyedLogger.Error(err, "Failed to get template from federated object")
+		return worker.StatusError
+	}
+
+	switch template.GroupVersionKind() {
+	case common.EndpointSliceGVK:
+		return c.reconcileEpsFedObj(ctx, template)
+	case common.ServiceImportGVK:
+		return c.reconcileSvcImportFedObj(ctx, fedObject, template)
+	}
+
+	return worker.StatusAllOK
+}
+
+func (c *ServiceImportController) reconcileEpsFedObj(
+	ctx context.Context,
+	unstructuredEps *unstructured.Unstructured,
+) (status worker.Result) {
+	logger := klog.FromContext(ctx)
+
+	svcName := unstructuredEps.GetLabels()[discoveryv1b1.LabelServiceName]
+
+	siFedObject, err := fedobjectadapters.GetFromLister(
+		c.fedObjectInformer.Lister(),
+		nil,
+		unstructuredEps.GetNamespace(),
+		naming.GenerateFederatedObjectName(svcName, serviceImportFtcName),
+	)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(2).Info("serviceImport federated object not found, no need to reconcile")
+			return worker.StatusAllOK
+		}
+
+		logger.Error(err, "Failed to get serviceImport federated object from store")
+		return worker.StatusError
+	}
+
+	serviceImportTemplate, err := siFedObject.GetSpec().GetTemplateAsUnstructured()
+	if err != nil {
+		logger.Error(err, "Failed to get serviceImport template from federated object")
+		return worker.StatusError
+	}
+	serviceImportObj := &mcsv1alpha1.ServiceImport{}
+	if err = runtime.DefaultUnstructuredConverter.FromUnstructured(serviceImportTemplate.UnstructuredContent(), serviceImportObj); err != nil {
+		logger.Error(err, "Failed to convert ServiceImport from unstructured object")
+		return worker.StatusError
+	}
+
+	if err := c.syncPlacementsToEndpointSlice(ctx, serviceImportObj,
+		siFedObject.GetSpec().GetControllerPlacement(PrefixedGlobalSchedulerName)); err != nil {
```

Comment:
Should we get scheduler placements or the union of all placements?

Reply:
Currently, serviceImport will only be scheduled by global-scheduler(follower-controller cannot). I think get scheduler placement is OK.